protocols/cosi/public_api.go

```diff
@@ -0,0 +1,459 @@
+	msg, _ := crypto.HashStream(network.Suite.Hash(), r)
+	req := &SignRequest{
+		EntityList: el,
+		Message:    msg,
+	}
+
+	// Connect to the root
+	host := el.List[0]
+	dbg.Lvl3("Opening connection to", host.Addresses[0], host.Public)
+	con, err := client.Open(host)
+	defer client.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	dbg.Lvl3("Sending sign SignRequest")
+	pchan := make(chan SignResponse)
+	go func() {
+		// send the SignRequest
+		if err := con.Send(context.TODO(), req); err != nil {
+			close(pchan)
+			return
+		}
+		dbg.Lvl3("Waiting for the response")
+		// wait for the response
+		packet, err := con.Receive(context.TODO())
+		if err != nil {
+			close(pchan)
+			return
+		}
+		pchan <- packet.Msg.(SignResponse)
+	}()
+	select {
+	case response, ok := <-pchan:
+		dbg.Lvl5("Response:", ok, response)
+		if !ok {
+			return nil, errors.New("Invalid repsonse: Could not cast the " +
+				"received response to the right type")
+		}
+		err = cosi.VerifySignature(network.Suite, msg, el.Aggregate,
+			response.Challenge, response.Response)
+		if err != nil {
+			return nil, err
+		}
+		return &response, nil
+	case <-time.After(time.Second * 10):
+		return nil, errors.New("Timeout on signing")
+	}
+}
+
+// verify takes a file and a group-definition, calls the signature
+// verification and prints the result
+func Verify(fileName, groupToml string) error {
+	// if the file hash matches the one in the signature
+	dbg.Lvl4("Reading file " + fileName)
+	b, err := ioutil.ReadFile(fileName)
+	if err != nil {
+		return err
+	}
+	// Read the JSON signature file
+	dbg.Lvl4("Reading signature")
+	sb, err := ioutil.ReadFile(fileName + ".sig")
+	if err != nil {
+		return err
+	}
+	sig := &SignResponse{}
+	dbg.Lvl4("Unmarshalling signature ")
+	if err := json.Unmarshal(sb, sig); err != nil {
+		return err
+	}
+	fGroup, err := os.Open(groupToml)
+	if err != nil {
+		return err
+	}
+	dbg.Lvl4("Reading group definition")
+	el, err := ReadGroupToml(fGroup)
+	if err != nil {
+		return err
+	}
+	dbg.Lvl4("Verfifying signature")
+	err = VerifySignatureHash(b, sig, el)
+	return err
+}
+
+// verifySignature checks whether the signature is valid
+func VerifySignatureHash(b []byte, sig *SignResponse, el *sda.EntityList) error {
+	// We have to hash twice, as the hash in the signature is the hash of the
+	// message sent to be signed
+	fHash, _ := crypto.HashBytes(network.Suite.Hash(), b)
+	hashHash, _ := crypto.HashBytes(network.Suite.Hash(), fHash)
+	if !bytes.Equal(hashHash, sig.Sum) {
+		return errors.New("You are trying to verify a signature " +
+			"belongig to another file. (The hash provided by the signature " +
+			"doesn't match with the hash of the file.)")
+	}
+	if err := cosi.VerifySignature(network.Suite, fHash, el.Aggregate, sig.Challenge, sig.Response); err != nil {
+		return errors.New("Invalid sig:" + err.Error())
+	}
+	return nil
+}
+
+// GroupToml represents the structure of the group.toml file given to the cli.
+type GroupToml struct {
```

Comment:
Same here, I think it should be placed under lib/cosi as it provide nice helpers.